samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -154,17 +149,13 @@
   private final Map<TaskName, Map<String, Set<SystemStreamPartition>>> taskSideInputStoreSSPs;
   private final Map<SystemStreamPartition, TaskSideInputHandler> sspSideInputHandlers;
   private SystemConsumers sideInputSystemConsumers;
-  private final Map<SystemStreamPartition, SystemStreamMetadata.SystemStreamPartitionMetadata> initialSideInputSSPMetadata
-      = new ConcurrentHashMap<>(); // Recorded sspMetadata of the taskSideInputSSPs recorded at start, used to determine when sideInputs are caughtup and container init can proceed
-  private volatile CountDownLatch sideInputsCaughtUp; // Used by the sideInput-read thread to signal to the main thread
+  private volatile Map<TaskName, CountDownLatch> sideInputTaskLatches; // Used by the sideInput-read thread to signal to the main thread
   private volatile boolean shouldShutdown = false;
+  private RunLoop sideInputRunLoop;
 
   private final ExecutorService sideInputsReadExecutor = Executors.newSingleThreadExecutor(
       new ThreadFactoryBuilder().setDaemon(true).setNameFormat(SIDEINPUTS_READ_THREAD_NAME).build());
 
-  private final ScheduledExecutorService sideInputsFlushExecutor = Executors.newSingleThreadScheduledExecutor(
-      new ThreadFactoryBuilder().setDaemon(true).setNameFormat(SIDEINPUTS_FLUSH_THREAD_NAME).build());
-  private ScheduledFuture sideInputsFlushFuture;
```

Comment:
so flush (aka commit) was happening in a separate thread right.. now we have moved that into the same thread as read (where the side input runloop now is). Why did we have two threads for read and flush earlier and what is the effect of removing the flush thread?

Reply:
@mynameborat @rmatharu I am not clear on the historical context here. Can one of you comment on why it was written this way?

From what I can tell, the change introduced to threading by my PR is that `commit` of one task will no longer be able to run concurrently with `process` of another task.

Reply:
CSM originally only handled state restoration which didn't require any commit/flush semantics. Eventually side inputs got rolled into CSM to pay way for standby and side inputs needed to be flushed in our commit cadence. The need for a separate timer thread to trigger flush/commit on task.commit.ms cadence was required. Instead of having logic to coordinate signals between read thread and the timer thread, flush was embedded as part of the timer thread and synchronization was introduced between process & commit.

With run loop, the internal timer thread run loop maintains does this job of signaling commit readiness to the task  and hence you don't need a separate flush thread. However, one thing to note is, previously we can potentially have two operations (process & flush) in parallel for different side inputs but with the new setup, we will need a thread pool size of (number of side input tasks + 1) to ensure parity. We will eventually support parallelism across stores/side input tasks as part of transaction state support (i guess). 

So it is fine to leave it as is.